Private Assets

dmitry-yudakov opened the pull request:
Original Asset3d functionality is now split into base Asset3d (options, meta) and UserAsset3d (user_id, meta, is_private). UserAsset3d keeps pointer to its base Asset3d.

API is mostly allowed to work with UserAsset3d. 
- adding - Asset3d is created if missing or existing one is used
- Edit/remove operations are allowed only for user's own objects. 
- GET operation reads user assets taking all public and current user's private assets; Meta of the response is merged from asset's base metadata and userAsset metadata.

Rest of the code uses asset3d as it used to.

- [x] UserAsset3d for keeping userID, meta and is_private for asset3d
- [x] Allow several assets with same asset_3d_id but different user_id
- [x] API
- [x] Seed data

dmitry-yudakov committed:
feat(asset_3d): support asset_3d_user, put userID/Private inside existing Asset3D, WIP

Potential direction of implementing the feature of private assets.

It will require keeping the assets in the map using concatenated asset3d_id and user_id.

dmitry-yudakov committed:
feat(assets-3d): use assetID-userID as unique key and use it everywhere, WIP

dmitry-yudakov committed:
Merge remote-tracking branch 'origin/develop' into feat/private-assets

dmitry-yudakov committed:
feat(assets-3d): better separation of asset3d and userAsset3d

dmitry-yudakov committed:
doc(readme): add info about local development with remote media-manager

jor-rit commented on README.md:
+42 bonus points for updating Readme! ;p

jor-rit commented on database/assets_3d/db.go:
```diff
@@ -82,22 +113,29 @@ func (db *DB) UpsertAssets(ctx context.Context, assets3d []*entry.Asset3d) error
+// 		return errors.WithMessage(err, "failed to exec db")
+// 	}
+// 	return nil
+// }
```
just delete the commented out func

jor-rit commented on universe/assets_3d/api.go:
```diff
@@ -18,14 +18,14 @@ func (a *Assets3d) RegisterAPI(r *gin.Engine) {
-			assets3d.GET("/meta", a.apiGetAssets3dMeta)
-			assets3d.GET("/options", a.apiGetAssets3dOptions)
+			// assets3d.GET("/meta", a.apiGetAssets3dMeta)
+			// assets3d.GET("/options", a.apiGetAssets3dOptions)
```
again, just delete 'em. We have git history if these ever need to come back :)

jor-rit commented on universe/assets_3d/assets_3d.go:
```diff
@@ -65,15 +68,67 @@ func (a *Assets3d) CreateAsset3d(asset3dID umid.UMID) (universe.Asset3d, error)
 
-func (a *Assets3d) FilterAssets3d(predicateFn universe.Assets3dFilterPredicateFn) map[umid.UMID]universe.Asset3d {
-	return a.assets.Filter(predicateFn)
+func (a *Assets3d) CreateAsset3dIfMissing(asset3dID umid.UMID) (universe.Asset3d, error, bool) {
```
Not a blocker, but I'd say just `CreateAsset3d` here as the public interface name, document the return bool as 'created' and users should get the intent.

dmitry-yudakov replied:
There is a simpler and more standard CreateAsset3d, hence CreateAsset3dIfMissing :)
But yeah, one of them is enough, so removed the old one and renamed the new one to CreateAsset3d.

jor-rit commented on universe/assets_3d/api_assets_3d.go:
```diff
@@ -42,32 +44,62 @@ func (a *Assets3d) apiGetAssets3d(c *gin.Context) {
+
+		asset3d := userAsset3d.GetAsset3d()
+		if asset3d == nil {
+			fmt.Printf("asset3d is nil, strange! %+v\n", userAsset3d.GetAssetUserIDPair())
```
log.Debugf

jor-rit commented on universe/assets_3d/api_assets_3d.go:
```diff
@@ -197,14 +187,28 @@ func (a *Assets3d) apiUploadAsset3d(c *gin.Context) {
 		return
 	}
 
+	fmt.Printf("Upload to mm response: %+v\n", response)
```
delete, or log.Debugf if you really really want it (but just use debugger breakpoint I'd suggest :)

dmitry-yudakov replied:
Yeah, sorry about these - some leftovers :)

dmitry-yudakov committed:
Merge remote-tracking branch 'origin/develop' into feat/private-assets

dmitry-yudakov committed:
refactor(assets-3d): remove obsolete functions

dmitry-yudakov committed:
refactor(assets-3d): better naming

dmitry-yudakov committed:
feat(assets-3d): restore /options EP - no other way to get this info

dmitry-yudakov committed:
refactor(assets-3d): leave only CreateAsset3dIfMissing, rename to CreateAsset3d

CreateAsset3dIfMissing is more complex function but it can be used everywhere.
Removed the other one and renamed to CreateAsset3dIfMissing.

dmitry-yudakov committed:
refactor(assets-3d): cleanup DB

dmitry-yudakov committed:
refactor(assets-3d): cleanup logs

dmitry-yudakov committed:
feat(assets-3d): change user asset is_private with existing patch EP

dmitry-yudakov committed:
fix(assets-3d): wrong mutex locked for userAssets operations

jor-rit commented on universe/user_asset_3d/user_asset_3d.go:
```diff
@@ -0,0 +1,141 @@
+
+func (a *UserAsset3d) IsPrivate() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
```
Should be a read lock, mu.RLock/RUnlock here

dmitry-yudakov committed:
fix(assets-3d): wrong mutex lock operation for reading Private

dmitry-yudakov committed:
feat(assets-3d): remove :objectID from API, BREAKING CHANGE

dmitry-yudakov committed:
refactor(assets-3d): better logs

dmitry-yudakov committed:
Merge remote-tracking branch 'origin/develop' into feat/private-assets

Update UserAsset3d to match the context changes

# Conflicts:
#	universe/asset_3d/asset_3d.go
#	universe/assets_3d/api_assets_3d.go
#	universe/assets_3d/assets_3d.go
